src/vc_util/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+pub const II_ISSUER_URL: &str = "https://internetcomputer.org/issuers/internet-identity";
+pub const VC_SIGNING_INPUT_DOMAIN: &[u8; 26] = b"iccs_verifiable_credential";
+
+/// A pair of identities, that denote the same user.
+/// Used in attribute sharing flow to maintain II's unlinkability of identities.
+pub struct AliasTuple {
+    /// A temporary identity, used in attribute sharing flow.
+    pub id_alias: Principal,
+    /// An identity under which a user is known to a dapp.
+    pub id_dapp: Principal,
+}
+
+#[derive(Debug)]
+pub enum CredentialVerificationError {
+    InvalidJws(SignatureVerificationError),
+    InvalidClaims(JwtValidationError),
+}
+
+/// Returns the effective bytes that will be signed when computing a canister signature for
+/// the given JWT-credential, verfiable via the specified public key.
+pub fn vc_signing_input(
+    credential_jwt: &str,
+    canister_sig_pk: &CanisterSigPublicKey,
+) -> Result<Vec<u8>, String> {
+    let encoder = jws_encoder(credential_jwt, canister_sig_pk)?;
+    Ok(encoder.signing_input().to_vec())
+}
+
+/// Computes and returns SHA-256 hash of the given `signing_input` prefixed with
+///      length(VC_SIGNING_INPUT_DOMAIN) || VC_SIGNING_INPUT_DOMAIN
+/// (for domain separation), where `length(a)` is the length of byte-array `a`,
+/// and || denotes concatenation of bytes.
+pub fn vc_signing_input_hash(signing_input: &[u8]) -> Hash {
+    let mut hasher = Sha256::new();
+    let buf = [VC_SIGNING_INPUT_DOMAIN.len() as u8];
+    hasher.update(buf);
+    hasher.update(VC_SIGNING_INPUT_DOMAIN);
+    hasher.update(signing_input);
+    hasher.finalize().into()
+}
+
+/// Constructs and returns a JWS (a signed JWT) from the given components.
+pub fn vc_jwt_to_jws(
+    credential_jwt: &str,
+    canister_sig_pk: &CanisterSigPublicKey,
+    sig: &[u8],
+) -> Result<String, String> {
+    let encoder = jws_encoder(credential_jwt, canister_sig_pk)?;
+    Ok(encoder.into_jws(sig))
+}
+
+/// Returns a DID for the given `principal`.
+pub fn did_for_principal(principal: Principal) -> String {
+    let prefix = String::from("did:icp:");
+    prefix.add(&principal.to_string())
+}
+
+/// Verifies the given JWS-credential as an id_alias-VC for the specified alias tuple.
+/// Performs both the cryptographic verification of the credential, and the semantic
+/// validation of the id_alias-claims.
+pub fn verify_id_alias_credential_jws(
+    credential_jws: &str,
+    alias_tuple: &AliasTuple,
+    root_pk_raw: &[u8],
+) -> Result<(), CredentialVerificationError> {
+    let claims = verify_credential_jws(credential_jws, root_pk_raw)
+        .map_err(CredentialVerificationError::InvalidJws)?;
+    validate_id_alias_claims(claims, alias_tuple)
+        .map_err(CredentialVerificationError::InvalidClaims)
+}
+
+/// Verifies the specified JWS credential cryptographically.
+/// DOES NOT perform semantic validation of the claims in the credential.
+pub fn verify_credential_jws(
```

Comment:
One important thing about canister signature verification is that you validate it with respect to a canister id. What about adding to this method another parameter (the issuer canister id) and also validate that the credential indeed was signed by the expected issuer?

Reply:
That's a good idea, done.
(I thought we'll have another helper for checking that, but I agree that this function is a good place for the check)